docs/proposals/20220414-runtime-hooks.md

```diff
@@ -0,0 +1,430 @@
+    * [Kubernetes version upgrade](#kubernetes-version-upgrade)
+* [Additional Details](#additional-details)
+    * [Test Plan](#test-plan)
+    * [Graduation Criteria](#graduation-criteria)
+    * [Version Skew Strategy](#version-skew-strategy)
+* [Implementation History](#implementation-history)
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+
+## Summary
+
+This proposal introduces a set of Runtime Hooks designed for allowing the implementation of add-on orchestration solutions on top of Cluster API.
+
+## Motivation
+
+Cluster Resource Set (CRS) is the current add-on management tool packaged with Cluster API, but many users rely on their own package management tool like helm, kapp, ArgoCD or flux, because those tools have a broad range of capabilities that are not currently available in CRS.
+
+The ability to orchestrate add-ons in line with events in the cluster lifecycle is becoming a requirement for many CAPI users, but in order to make this possible a mechanism to plug into the cluster lifecycle is required. This proposal introduces a set of Runtime Hooks designed to meet the need for add-on management orchestration including:
+
+* Operations for installing add-ons during the cluster provisioning workflow
+* Operations for upgrading add-ons during the cluster upgrade workflow
+* Operations for handling add-ons during the cluster deletion workflow
+
+Runtime Hooks enable the Cluster Lifecycle to trigger these processes based on the current state of the cluster, allowing them to start after some state has been reached or waiting for them to complete before moving on with Cluster-wide operations such as Cluster creation or deletion.
+
+Once these hooks are in place, it will be possible to build a comprehensive add-on orchestration solution on top of Cluster API that can leverage external tools such as helm, kapp, ArgoCD, flux or eventually CRS as an alternative.
+
+The hooks defined in this proposal are going to model common events in the Cluster lifecycle and we assume that in the future they may be used for other use cases beyond add-on management. 
+
+### Goals
+
+* Identify a set of Runtime Hooks that enable management of the entire add-on lifecycle 
+* Define the OpenAPI specification of these Runtime Hooks
+* Document when the corresponding Runtime Extensions are called
+* Provide guidelines for developers implementing a corresponding Runtime Extension
+
+### Non-Goals
+
+* Defining all possible Runtime Hooks in Cluster API; this proposal defines only a subset of hooks required for add-on orchestration.
+* Define a full add-on management solution or define detailed steps for solving add-on related problems like CPI migration from in-tree to out-of-tree; this proposal is focused only in providing the foundational capabilities to do so. 
+
+### Future work
+
+* Improve the hooks defined in this proposal to better support use cases beyond add-on management 
```

Comment:
Just to clarify, does this mean introducing new hooks or modify the existing ones?

Reply:
I think new hooks are outside of the scope of this proposal (unless a new hook is needed for add-on management). This is designed to cover over any changes that need to be made as we understand the add-on story and how suitable these hooks are and may involve e.g. changing the response / request types.

If we need a new hook specifically for add-on management (hopefully not :slightly_smiling_face: ) that could be part of the future work in the proposal - other hooks should have their own proposals / PRs IMO

Reply:
The current sentence reads like future work is to improve the hooks **currently** defined in this proposal so they are also cover uses-cases apart from add-on management.

Let's try to rephrase slightly so it's 100% clear what we mean.

Reply:
If we can rephrase ok, but IMO we can also remove this given that it is kind of implicit that this feature can be improved in the future if other use cases add more requirements for this set of runtime hooks

Reply:
I am +1 to dropping this section.